packages/falter-berlin-service-registrar/files/register-services.sh

```diff
@@ -0,0 +1,241 @@
+# Copyright (C) 2022   Martin Hübner
+
+# shellcheck source=/dev/null
+. /lib/functions.sh
+
+# ToDo: Website descriptions must not contain a comma!
+
+
+log() {
+    local msg="$1"
+    logger -t "ffserviced" -s "$msg"
+}
+
+# busybox doesn't include the rev-command. Thus implementing it by ourself.
+rev() {
+    while read -r var; do
+        # function taken from https://stackoverflow.com/a/34668251
+        rev=""
+        i=1
+
+        while [ "$i" -le "${#var}" ]; do
+            rev="$(echo "$var" | awk -v i="$i" '{print(substr($0,i,1))}')$rev"
+            : $((i += 1))
+        done
+
+        echo "$rev"
+    done
+}
+
+get_olsrd_nsplugin_index() {
+    # get anonymous section number of nameservice_plugin in olsrd-config
+    local ipversion="$1"
+
+    if [ "$ipversion" = 4 ]; then
+        uci show olsrd | grep nameservice | sed -e 's|.*\(\d\).*|\1|g'
+    elif [ "$ipversion" = 6 ]; then
+        uci show olsrd6 | grep nameservice | sed -e 's|.*\(\d\).*|\1|g'
+    else
+        log "cannot find olsrd-conf for IP-Version $ipversion."
+    fi
+}
+
+flush_nondefault_uhttpd_sections() {
+    local section_name="$1"
+
+    if [ "$section_name" != "main" ] && [ "$section_name" != "defaults" ]; then
+        # flush section
+        uci_remove uhttpd "$section_name"
+    fi
+}
+
+flush_olsrd_services() {
+    nsplugin=$(get_olsrd_nsplugin_index 4)
+
+    uci_remove olsrd @LoadPlugin["$nsplugin"] hosts
+    uci_remove olsrd @LoadPlugin["$nsplugin"] service
+}
+
+# This script will empty all entries made in the olsrd nameservice plugin
+# So only those entries will remain, that are handled by this script,
+flush_old_entries() {
+    config_load uhttpd
+    config_foreach flush_nondefault_uhttpd_sections
+
+    flush_olsrd_services
+}
+
+create_website_template() {
+    local path="$1"
+
+    mkdir -p "$path" || return 1
+    if [ ! -f "$path"/index.html ]; then
+        cat <<EOF >"$path"/index.html
+<!doctype html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8">
+    <title>It works!</title>
+  </head>
+  <body>
+    <h1>It works!</h1>
+    <p>You've successfully registered your static web page as a service in the Freifunk Berlin network. You may replace this example page with your own static website now.</p>
+    <p>This file is located under the service root directory at $path/index.html</p>
+  </body>
+</html>
+EOF
+        if [ $? != 0 ]; then
+            return 1
+        fi
+    fi
+}
+
+register_website_at_uhttpd() {
+    # configure website in uhttpd
+    local uci_section="$1"
+    local fqdn="$2"
+    local port="$3"
+    local web_root="$4"
+
+    uci_add uhttpd uhttpd "$uci_section"
+    uci_add_list uhttpd "$uci_section" listen_http "0.0.0.0:$port"
+    uci_add_list uhttpd "$uci_section" listen_http "[::]:$port"
+    # ToDo: Test, that uci handles "/tmp/www/website//" correctly
+    uci_set uhttpd "$uci_section" home "$web_root/"
+    uci_set uhttpd "$uci_section" max_requests 5
```

Comment:
are those good defaults? Can you benchmark this does not harm the node with "ab - Apache HTTP server benchmarking tool"?

Reply:
Testing on my Archer C5 at home gives this results:

```sh
ab -t 30 -n 10000 -c 100 wh.pcfw.eu:81/
This is ApacheBench, Version 2.3 <$Revision: 1903618 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking wh.pcfw.eu (be patient)
Finished 784 requests


Server Software:        
Server Hostname:        CENSORED
Server Port:            81

Document Path:          /
Document Length:        427 bytes

Concurrency Level:      100
Time taken for tests:   30.003 seconds
Complete requests:      784
Failed requests:        0
Total transferred:      487648 bytes
HTML transferred:       334768 bytes
Requests per second:    26.13 [#/sec] (mean)
Time per request:       3826.877 [ms] (mean)
Time per request:       38.269 [ms] (mean, across all concurrent requests)
Transfer rate:          15.87 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:       15  781 2027.7     21   15405
Processing:    15   22   9.9     20      91
Waiting:       14   22   9.5     20      90
Total:         31  803 2027.7     41   15426

Percentage of the requests served within a certain time (ms)
  50%     41
  66%     47
  75%    104
  80%   1054
  90%   3107
  95%   3171
  98%   7164
  99%   7176
 100%  15426 (longest request)
 ```
 
This was tested from the internet, not on the local LAN.

Reply:
The question is what does the load of the node say. Is it still able to perform it's normal tasks?

Reply:
Okay, I did perform another test from my local network:

```
ab -t 60 -n 100000 -c 1000 frei.funk:81/
This is ApacheBench, Version 2.3 <$Revision: 1903618 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking frei.funk (be patient)
Finished 1544 requests


Server Software:        
Server Hostname:        frei.funk
Server Port:            81

Document Path:          /
Document Length:        427 bytes

Concurrency Level:      1000
Time taken for tests:   60.745 seconds
Complete requests:      1544
Failed requests:        0
Total transferred:      960368 bytes
HTML transferred:       659288 bytes
Requests per second:    25.42 [#/sec] (mean)
Time per request:       39342.756 [ms] (mean)
Time per request:       39.343 [ms] (mean, across all concurrent requests)
Transfer rate:          15.44 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        2 1184 2885.2      5   31932
Processing:     3   13  20.5      6     142
Waiting:        2   12  20.4      5     141
Total:          6 1197 2885.7     15   31935

Percentage of the requests served within a certain time (ms)
  50%     15
  66%   1016
  75%   1038
  80%   1061
  90%   3119
  95%   7132
  98%   7160
  99%  15341
 100%  31935 (longest request)
```
There is a notably spike in the cpu usage, but I think that is okay:

```
~# uptime
 12:02:07 up 1 day,  9:16,  load average: 0.41, 0.23, 0.08
```

When the test was running I was able to use LuCI fine. I didn't measure it in comparison to a no load scenario, but getting the pages happend in under 2 seconds, which is almost the same it takes in no load too.

Do you have some specific thing in mind, like allowing even less requests or so?

Reply:
no, this is perfect. just wanted to make sure there is no negative effect.